discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,115 @@
+package zookeeper
+
+import (
+	"path"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/swarm/discovery"
+	"github.com/samuel/go-zookeeper/zk"
+)
+
+type ZkDiscoveryService struct {
+	conn      *zk.Conn
+	path      string
+	heartbeat int
+}
+
+func init() {
+	discovery.Register("zk", &ZkDiscoveryService{})
+}
+
+func (s *ZkDiscoveryService) Initialize(uris string, heartbeat int) error {
+	var (
+		// split here because uris can contain multiples ips
+		// like `zk://192.168.0.1,192.168.0.2,192.168.0.3/path`
+		parts = strings.SplitN(uris, "/", 2)
+		ips   = strings.Split(parts[0], ",")
+	)
+
+	conn, _, err := zk.Connect(ips, time.Second)
+
+	if err != nil {
+		return err
+	}
+
+	s.conn = conn
+	s.path = "/" + parts[1]
+	s.heartbeat = heartbeat
+
+	_, err = conn.Create(s.path, []byte{1}, 0, zk.WorldACL(zk.PermAll))
+	if err != nil {
+		// if key already existed, then skip
+		if err != zk.ErrNodeExists {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *ZkDiscoveryService) Fetch() ([]*discovery.Node, error) {
+	addrs, _, err := s.conn.Children(s.path)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var nodes []*discovery.Node
+	for _, addr := range addrs {
+		nodes = append(nodes, discovery.NewNode(addr))
+	}
+
+	return nodes, nil
+}
+
+func (s *ZkDiscoveryService) Watch(callback discovery.WatchCallback) {
+	for _ = range time.Tick(time.Duration(s.heartbeat) * time.Second) {
```

Comment:
Can you use zookeeper watches instead of a timer here please ?

Like we do in etcd.